### PR TITLE
tools: Add ChangeLog helper script

### PR DESCRIPTION
**Pull request by hoffie**
**Short description of changes**

This PR adds release tooling to validate the ChangeLog for completeness.
- find-missing-entries: Will find missing PRs/Issues by scanning the Github milestone and the `git log`.
- add-missing-entries: Will edit the ChangeLog and add those found entries.
- group-entries: Will group newly added (or even existing) entries by their category while using a pre-defined category order (within the script).

**Context: Fixes an issue?**

**Does this change need documentation? What needs to be documented and how?**

- [x] The `CHANGELOG:` syntax should be suggested in the PR template. This is done as part of this PR.
- [x] Translation logic still has whitespace issues
- [ ] Existence of this tool should be documented in the release checklist: https://github.com/jamulussoftware/jamuluswebsite/pull/708

**Status of this Pull Request**

Works, but open to improvements.

**What is missing until this pull request can be merged?**

Reviews.

## Checklist

- [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [x] I tested my code and it does what I want
- [ ] ~My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency)~  (no bash style guide)
- [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [x] I've filled all the content above

---

> Why is #1993 listed here? It was tagged for 3.8.1, not 3.8.2, but was actually closed, not merged, and replaced with #1994. Just thought it might be something to check in your tools.

_Originally posted by @softins in https://github.com/jamulussoftware/jamulus/issues/2249#issuecomment-1017745618_

It's listed because it's part of the history in master (see a8853b6a7deb19e3596fb47bbb72a5baf29c4c7a). Might be a stray reference.
The tool uses Github PRs and their milestone to find relevant PRs. However, we might have forgotten to assign a milestone. That's why the script also scans `git log` and uses the references there as well.

Finally, let this PR be handled properly in the tool...:
CHANGELOG: SKIP

### CODE REVIEW

**Review by softins — COMMENTED** [review]
When trying this out, I was in a working directory that had my own fork as the base repo for `gh` (in `.git/config`):
```
[remote "origin"]
    url = git@github.com:softins/jamulus.git
    fetch = +refs/heads/*:refs/remotes/origin/*
    gh-resolved = base
```
That meant it couldn't view any of the upstream PRs:
```
tony@pi:~/jamulus $ gh pr list

There are no open pull requests in softins/jamulus
```
If I switched to my other working directory that only had `jamulussoftware/jamulus` as a remote, `gh pr` worked fine:
```
tony@pi:~/jamulus-upstream $ gh pr list

Showing 20 of 20 open pull requests in jamulussoftware/jamulus
......
```
One suggestion would be to replace `gh` with `$GH` everywhere, and then at the top do:
```
GH="gh -R jamulussoftware/jamulus"
```
It would then look at the upstream PRs whatever remote the user had set as the `gh` base.

Alternatively, prominently document the need to set `gh-resolved` to the upstream repo if running multiple remotes, and/or abort the script with an explanation if `gh pr view N` returns non-zero.

Once I understood the problem, it was easy to change the `gh` base in my normal working directory, to point to the `upstream` repo by default.

File: tools/changelog-helper.sh [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        -re 's/\s*\.?\s*$//' \
+        -re 's/\b((Add)|(Updat|Enhanc|Improv|Remove)e)\b/\2\3ed/i'
+}
+
+case "${1:-1}" in
+    find-missing-entries)
+        ACTION=find-missing-entries
+        ;;
+    add-missing-entries)
+        ACTION=add-entries
```
[the remark this answers is not in the document]
Comment (hoffie) [reply tools/changelog-helper.sh]: Thanks, fixed.

File: tools/changelog-helper.sh [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    echo "Checking if all PR or references in git log since ${prev_release_tag} are included for ${target_release} based on ref ${target_ref}..."
+    for id in $(git log ${prev_release_tag}..master | grep -oP '#\K(\d+)'); do
+        gh pr view ${id} --json title &>/dev/null || continue # Skip non-PRs
+        check_or_add_pr "${id}"
+    done
+
+    echo
+    echo "Done."
+    if [[ $ACTION == find-missing-entries ]]; then
+        echo "You can re-run this script with add-entries to fill the ChangeLog automatically."
```
[the remark this answers is not in the document]
Comment (hoffie) [reply tools/changelog-helper.sh]: Thanks, fixed.

**Comment by hoffie** [comment]
> I was in a working directory that had my own fork as the base repo for gh

Ah, didn't think of that, but that's certainly important to handle properly.

> ```
> export GH_REPO=jamulussoftware/jamulus
> ```

Sounds good, I've added that.

File: tools/changelog-helper.sh [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    fi
+    echo ", adding new entry"
+    local title_suggestion_in_pr=$(gh pr view "$id" --json body,comments | grep -oP '\\n\s*CHANGELOG:\s*\K([^\\]{5,})' | tail -n1 | sanitize_title)
+    if [[ ${title_suggestion_in_pr} ]]; then
+        title=${title_suggestion_in_pr}
+        echo  "   using CHANGELOG suggestion from PR -> ${title}"
+        if [[ ${title_suggestion_in_pr} == "SKIP" ]]; then
+            continue
+        fi
+    fi
```
[the remark this answers is not in the document]
Comment (hoffie) [reply tools/changelog-helper.sh:123]: Hrm, I think it already should (unless I misunderstand)?

The current logic should be like this:

- If there is a `CHANGELOG:` line in the PR, it is interpreted.
- If the line equals `SKIP`, the whole entry is skipped. There will be no placeholder either. This is intended for PRs like this one which do not necessarily have to be part of the ChangeLog.
- If the line is not `SKIP`, it is used verbatim.
- If there's not `CHANGELOG:` line, the PR title is used.

This probably implies that we could do us a favor by editing PR titles and/or adding `CHANGELOG:` comments on merge.

File: tools/changelog-helper.sh [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        "Client"
+        "GUI"
+        "$TRANSLATION_ENTRY_TEXT"
+        "Server"
+        "Recorder"
+        "Performance"
+        "CLI"
+        "Bug Fix"
+        "Windows"
+        "Installer"
```
[the remark this answers is not in the document]
Comment (hoffie) [reply tools/changelog-helper.sh:88]: We didn't have any changes in this area for some time, but I wouldn't rule out we may have some in the future again? as such, I'd keep this.

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -1,7 +1,9 @@
 <!-- Thank you for working on Jamulus and opening a Pull Request! Please fill the following to make the review process straight forward -->
 
 **Short description of changes**
-<!-- A short description of your changes which might go into the change log -->
+<!-- Explain what your PR does -->
+
+CHANGELOG: <!-- Insert a short, end-user understandable sentence in past tense right here, e.g.: Client: Fixed crash when clicking too fast -->
```
Comment (ann0see) [inline .github/pull_request_template.md:6]:
```suggestion
CHANGELOG: <!-- Insert a short, end-user understandable sentence in past tense right here, e.g.: Client: Fixed crash when clicking the connect button too fast -->
```

It should also be precise

**Review by softins — APPROVED** [review]
Looks good to me.

File: .github/pull_request_template.md (hunk 5 above)
Comment (hoffie) [reply .github/pull_request_template.md]: Changed, thanks.

**Comment by pljones** [comment]
Is this aiming for 3.9.0?  Should it have a milestone set?

File: tools/changelog-helper.sh [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    echo
+    echo "Checking if all PR or references in git log since ${prev_release_tag} are included for ${target_release} based on ref ${target_ref}..."
+    for id in $(git log ${prev_release_tag}..master | grep -oP '#\K(\d+)'); do
+        gh pr view ${id} --json title &>/dev/null || continue # Skip non-PRs
+        check_or_add_pr "${id}"
+    done
+
+    echo
+    echo "Done."
+    if [[ $ACTION == find-missing-entries ]]; then
```
Comment (ann0see) [inline tools/changelog-helper.sh:54]: Shouldn’t we quote all strings?

File: tools/changelog-helper.sh [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        echo "You can re-run this script with add-missing-entries to fill the ChangeLog automatically."
+    fi
+}
+
+group_entries() {
+    local changelog_begin_position=$(grep -nP '^### .*\d+\.\d+\.\d+\b' ChangeLog | head -n1 | cut -d: -f1)
+    local changelog_prev_release_position=$(grep -nP '^### .*\d+\.\d+\.\d+\b' ChangeLog | head -n2 | tail -n1 | cut -d: -f1)
+
+    # Save everything before the actual release changelog content:
+    local changelog_header=$(head -n $changelog_begin_position ChangeLog)
```
Comment (ann0see) [inline tools/changelog-helper.sh:64]: Quoting

File: tools/changelog-helper.sh [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+            # No existing translation entry at all. Add a new one.
+            changelog="${changelog}$(
+                echo
+                echo
+                echo "- ${TRANSLATION_ENTRY_TEXT}"
+                echo "  * ${full_lang}, by @${author} (#${id})"
+			)"
+        else
+            # Existing translation entries, so sort them:
+            local changelog_before_translations=""
```
Comment (ann0see) [inline tools/changelog-helper.sh:198]: Can sed edit "in-place" so we don‘t need to copy loads of data into a variable?

File: tools/changelog-helper.sh [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    # Rebuild the changelog and write back to file:
+    (echo "$changelog_header"; echo "$changelog"; echo; echo "$changelog_prev_releases") > ChangeLog
+}
+
+sanitize_title() {
+    sed \
+        -re 's/^\s+//' \
+        -re 's/\s{2,}/ /' \
+        -re 's/\s*\.?\s*$//' \
+        -re 's/\b((Add)|(Updat|Enhanc|Improv|Remove)e)\b/\2\3ed/i'
```
Comment (ann0see) [inline tools/changelog-helper.sh:240]: Why is Remove not Remov here?

I didn’t fully analyse what this line does

**Comment by ann0see** [comment]
After thinking and looking at the past change logs, it seems as if we got shorter in our description style of changes. Is this good or bad?
https://github.com/jamulussoftware/jamulus/blob/8bbff18900d720317ad2c98c2cb1c225d6abbc00/ChangeLog#L62
for example is not that verbose

https://github.com/jamulussoftware/jamulus/blob/8bbff18900d720317ad2c98c2cb1c225d6abbc00/ChangeLog#L260-L263

For example gives a lot more detail.

We could still give users a little advice how to use the new feature (and not just give a commit-message style entry).

**Comment by pljones** [comment]
The change log entries shouldn't be limited to a single line, for sure.  In my view, they shouldn't, however, attempt to explain the problem and what was done to solve it -- only to say what has improved and link to the full description (i.e. issue, if available, and pull request).  So anything saying "Improve <something>" is a bit redundant :).

> Improved the version output on the command line to give more detail and Qt version

could be

> Version description now provides more detail and Qt version

As it covers all cases (GUI, non-GUI, client, server) adding a limiting clause is redundant.  However, "more detail" doesn't really say anything, either...  What changed? :)

**Comment by hoffie** [comment]
> Section, if it's a singular limiting feature. If multiple areas are affected, don't use the "tag:" format, use the content if it's necessary to limit the scope of the description - it may be obvious.

I really like the tag-style as it makes it possible to group entries and to enable Changelog skimming. As a Linux user I could skip all Windows entries. I agree thought hat there are types of changes which are hard to tag, so there may be exemptions.

> Source: the issue (if raised) and PR

I think we haven't been consistenly adding the issue numbers as well. The changelog helper currently doesn't (but it probably could when the Github metadata is up-to-date (`Fixes:`))

> I don't think "contributed by" needs saying -- the format is condensed: the text in brackets at the end is a list (separated by commas, no "and") of contributors. If that needs explaining, say it once.

I've just taken the style which has been there (I think it moved from `coded by` to `contributed by` which is a bit more generic), but I'm open to changing that.

In general, I'd rather like to get this PR merged at some point. It's been sitting around for quite some time, yet it is the tool I've been using to generate all the recent changelog PRs. We can always iterate on further improvement ideas later.

File: tools/changelog-helper.sh (hunk 9 above)
Comment (hoffie) [reply tools/changelog-helper.sh:240]: Thanks, fixed.

File: tools/changelog-helper.sh (hunk 7 above)
Comment (hoffie) [reply tools/changelog-helper.sh:64]: Fixed, thanks.

File: tools/changelog-helper.sh (hunk 6 above)
Comment (hoffie) [reply tools/changelog-helper.sh:54]: Fixed, thanks.

File: tools/changelog-helper.sh (hunk 8 above)
Comment (hoffie) [reply tools/changelog-helper.sh:198]: `sed` does support in-place editing, but I don't think those features are sufficient to do what's needed here. Feel free to suggest an alternative, that's what I came up with in absence of any better approaches.

**Comment by pljones** [comment]
> > Section, if it's a singular limiting feature. If multiple areas are affected, don't use the "tag:" format, use the content if it's necessary to limit the scope of the description - it may be obvious.
> 
> I really like the tag-style as it makes it possible to group entries and to enable Changelog skimming. As a Linux user I could skip all Windows entries. I agree thought hat there are types of changes which are hard to tag, so there may be exemptions.

Yes, by "Section" I mean "UI:" or whatever.  But "UI, Command Line, Client, Server:" seems a bit unhelpful... it's really helpful if the change only affects one area and becomes less helpful the more areas the change covers.

File: tools/changelog-helper.sh (hunk 8 above)
Comment (ann0see) [reply tools/changelog-helper.sh:198]: Maybe include a FIXME comment explaining that this could be improved

**Review by ann0see — APPROVED** [review]
Going to approve this now.

**Pull request merged**